### ISSUE TITLE
[Feature]: Autocorrect / gesture recognition training game

### ISSUE THREAD

**Issue by BurntVoxel** [tribixbite/CleverKeys#101]
### Feature Category

Neural Prediction (AI models, accuracy, speed)

### Feature Description

If autocorrect is powered by a neural network, what if we could deliberately train it in a "safe space" rather than only while we're actually trying to use it? Another FOSS keyboard project by FUTO tried this, though theirs was a web page that collected training data. We could make use of the same idea here but built into the app! 

Essentially the idea is to generate network training data by having the user read text and retype it, "show me what the gestures for these words look like and what words are more common."

### Problem This Solves

- Autocorrect always kinda sucks at first, no matter who developed it, even Google's keyboard. 
- Breaking autocorrect in faster helps retain users. (Personal anecdote: I tried the app, couldn't daily-drive it because another one was already trained, forgot about it, keep finding it and saying I want to try again. It's cool, it's just the autocorrect.)
- Better training data: you're not guessing anything, you have confidence you know what they meant to type.
- Also offers a little activity that's fun but productive: improve your keyboard experience while reading some articles

### Proposed Solution

- Copy source text from a URL or a given file or copy/paste. (Maybe suggest Wikipedia's random article button or Project Gutenberg books)
- Show the source text as written, dimmed. Ask the user to type it verbatim. As they type, color the active word brightly and the past text slightly less
- Tweak backspace to do whole word deletion for now
- Log every detail. Every typo. Every gesture. Every bad attempt and backspace and redo. 
- Know that this data is valuable for ALL kinds of mistakes. Physical typos like mashing the wrong key, failing at spelling, and thinking of the wrong word. In all those cases, you can actually see what they meant.
- Even if they read the word wrong, it's still good training data. Take etymology/entomology mixup: you still see the words around it that provide context clues, the word we meant, and the word we said by mistake.
- Machine learning...?
- profit

### Alternative Solutions

- Standalone app, crowdsource data for the starter model
- Rather than tweak autocorrect, tweak the config file numbers until simulating the same log gets the best autocorrect results.

### Priority Level

Nice to have (would be cool but not essential)

### Implementation Considerations

- [x] This might require neural model changes
- [x] This might need new UI components
- [ ] This could impact performance
- [ ] This might affect privacy/security
- [ ] This could require system-level permissions
- [ ] This might need hardware acceleration
- [ ] I'm willing to help implement this

### Neural/AI Considerations (if applicable)

This would provide more data for the models. The production side would probably not need changed, but maybe the training side would. 

### Privacy Considerations

- Assuming you don't crowdsource the data for the starter model, data shouldn't need to leave the phone... long as you train when plugged into a charger.
- Assuming you do crowdsource, it should be made clear up front. At least you're training on a public domain book instead of someone's chat logs

### Additional Context

Perhaps check out Futo's implementation, or request their dataset if you haven't. https://swipe.futo.org/

**Comment by session11** [comment]
I don't personally use swipe typing (old school) but the auto correct in cleverkeys lacks efficiency (tap method typing) . I switched from Multilingo Keyboard (another great keyboard with powerful gestures and complete DIY concept) where typing gives you a lot of confidence. No matter how fuzzy I type, the keyboard almost always corrects the typos.  Typing with cleverkeys forces me to pay attention to each tapped key. One mistake and the intended word is gone. For example I usually mistype "tge", tfe" and so on for the intended word "the" but cleverkeys never auto corrects it no matter how much I fiddle with the settings. Another example: if I wish to type "question" and mistype just one word as "wuestion", it will never auto correct it and instead asks me to save the wrong word. In short, the experience is frustrating at times. I even tried to Swipe type but the irony is I never managed to get the word "SWIPE" with swiping across the keyboard.

**Comment by tribixbite** [comment]
> I don't personally use swipe typing (old school) but the auto correct in cleverkeys lacks efficiency (tap method typing) . I switched from Multilingo Keyboard (another great keyboard with powerful gestures and complete DIY concept) where typing gives you a lot of confidence. No matter how fuzzy I type, the keyboard almost always corrects the typos. Typing with cleverkeys forces me to pay attention to each tapped key. One mistake and the intended word is gone. For example I usually mistype "tge", tfe" and so on for the intended word "the" but cleverkeys never auto corrects it no matter how much I fiddle with the settings. Another example: if I wish to type "question" and mistype just one word as "wuestion", it will never auto correct it and instead asks me to save the wrong word. In short, the experience is frustrating at times. I even tried to Swipe type but the irony is I never managed to get the word "SWIPE" with swiping across the keyboard.

Those are bugs (and regressions) - I will fix them and make autocorrect way better. 

Eventually game/activity to improve sure

**Comment by BurntVoxel** [comment]
> Those are bugs (and regressions) - I will fix them and make autocorrect way better.
> 
> Eventually game/activity to improve surr

Forgive me for asking before getting a good look at the codebase, but, have you considered making autocorrect consider the physical proximity of the keys when calculating difference between two words? (Idk if there's a proper term for this, maybe key adjacency? But I'll call it typo distance) 

Examples:
- If I mean `you` but write `yiu`, the word-difference is 1 letter substitution, one key-width away. Autocorrect may think I meant `toy`, but the difference would be 3 substitutions, 3 distance, so it should be ranked lower. 
- `nat -> net`: on querty the center of E is about one and a half keys away from the center of A, so the difference should be `1Sub + 1.5Dist`
- `medt -> ?`: `neat` is difference `1sub + 1dist + 1.5dist`, `nest` is `1sub + 1dist + 1dist`, `meat` is `1sub + 2dist`, `meet` is `1sub + 1dist`... 

Of course, how much it should weigh each type of error should be configurable. And there should be other metrics, like, is a letter substitution a phonics problem, like "is that an E sound or an I sound", and so on. 


... Which brings me to something I didn't put enough thought into when writing the original post. 
> Alternative solution: Rather than tweak autocorrect, tweak the config file numbers until simulating the same log gets the best autocorrect results.

There's a *lot* of parameters that screw with how well we can type. Keyboard spacing, autocorrect confidence threshold, gesture distance threshold, minimum word frequency, rare word penalty, correction search depth...

Plus the difference probably depends on the individual. Like, doctors use rare words a lot. Personally I'm much more likely to do a typo like hitting the adjacent key than to make an actual spelling mistake, but some people do actually make such mistakes. 

Getting all the knobs and dials right is hard, but... Perhaps we could use a similar algorithm to machine learning to adjust all the settings (or, all the settings which actually impact your typing ability or autocorrect quality). Which is where my proposal could be repurposed: get sample text paired with user input, and "train" the config, tweak it until it interprets the same raw input as similarly to the sample text as possible. 

I don't know much about the neural network behind autocorrect, like if it can be retrained or if the model is frozen, but the settings parameters feel like they should have a much more direct impact on quality, since they control the accuracy of the input to begin with. Plus there's much fewer parameters, yet still so many that tuning by hand feels difficult. And there's no worry about it being locked in, just export a backup of the config first.

**Comment by session11** [comment]
I will be eagerly waiting for the enhanced autocorrect feature. 🤞

**Comment by tribixbite** [comment]
> I will be eagerly waiting for the enhanced autocorrect feature. 🤞

If you have interest in testing, the latest dev build has major autocorrect architectural changes. There were several bugs with frequency scaling and char match preset.

It also weights candidates by key proximity

**Comment by tribixbite** [comment]
> ### Feature Category
> 
> Neural Prediction (AI models, accuracy, speed)
> 
> ### Feature Description
> 
> If autocorrect is powered by a neural network, what if we could deliberately train it in a "safe space" rather than only while we're actually trying to use it? Another FOSS keyboard project by FUTO tried this, though theirs was a web page that collected training data. We could make use of the same idea here but built into the app! 
> 
> Essentially the idea is to generate network training data by having the user read text and retype it, "show me what the gestures for these words look like and what words are more common."
> 
> ### Problem This Solves
> 
> - Autocorrect always kinda sucks at first, no matter who developed it, even Google's keyboard. 
> - Breaking autocorrect in faster helps retain users. (Personal anecdote: I tried the app, couldn't daily-drive it because another one was already trained, forgot about it, keep finding it and saying I want to try again. It's cool, it's just the autocorrect.)
> - Better training data: you're not guessing anything, you have confidence you know what they meant to type.
> - Also offers a little activity that's fun but productive: improve your keyboard experience while reading some articles
> 
> ### Proposed Solution
> 
> - Copy source text from a URL or a given file or copy/paste. (Maybe suggest Wikipedia's random article button or Project Gutenberg books)
> - Show the source text as written, dimmed. Ask the user to type it verbatim. As they type, color the active word brightly and the past text slightly less
> - Tweak backspace to do whole word deletion for now
> - Log every detail. Every typo. Every gesture. Every bad attempt and backspace and redo. 
> - Know that this data is valuable for ALL kinds of mistakes. Physical typos like mashing the wrong key, failing at spelling, and thinking of the wrong word. In all those cases, you can actually see what they meant.
> - Even if they read the word wrong, it's still good training data. Take etymology/entomology mixup: you still see the words around it that provide context clues, the word we meant, and the word we said by mistake.
> - Machine learning...?
> - profit
> 
> ### Alternative Solutions
> 
> - Standalone app, crowdsource data for the starter model
> - Rather than tweak autocorrect, tweak the config file numbers until simulating the same log gets the best autocorrect results.
> 
> ### Priority Level
> 
> Nice to have (would be cool but not essential)
> 
> ### Implementation Considerations
> 
> - [x] This might require neural model changes
> - [x] This might need new UI components
> - [ ] This could impact performance
> - [ ] This might affect privacy/security
> - [ ] This could require system-level permissions
> - [ ] This might need hardware acceleration
> - [ ] I'm willing to help implement this
> 
> ### Neural/AI Considerations (if applicable)
> 
> This would provide more data for the models. The production side would probably not need changed, but maybe the training side would. 
> 
> ### Privacy Considerations
> 
> - Assuming you don't crowdsource the data for the starter model, data shouldn't need to leave the phone... long as you train when plugged into a charger.
> - Assuming you do crowdsource, it should be made clear up front. At least you're training on a public domain book instead of someone's chat logs
> 
> ### Additional Context
> 
> Perhaps check out Futo's implementation, or request their dataset if you haven't. https://swipe.futo.org/

Futo's credited as a dataset source in the readme. 

There's actually already a mini game of sorts that's been in the app since v1 (Advanced-> Swipe Calibration) - mostly played by me as I tried to beat my previous accuracy + ms latency score for each app iteration, but I also hoped to have it be part of an on-device user tuning pipeline similar to what you've outlined. 

However being a new privacy focused app made it challenging to do both (have local training (= "tracking" in peoples eyes) and be privacy focused). Many people don't care /want to hear the distinction and, for now, it's just simpler to push back on device learning.

There's plenty of room for major improvement in swipe and autocorrect accuracy and user config, so it makes sense to start there.

For outsourced data gathering, I'm most interested in collecting samples from people using alt (non qwerty) layouts. But even higher priority (ranked by request frequency) is multi-lang swipe engine.... and getting more users will make it much easier to facilitate building a useable dataset - so tldr that's step 1!

See "The Key Proximity System" https://cleverkeys.app/wiki/typing/autocorrect/
(Added and built today)
